MINOR: A bunch of clean-ups related to usage of unused variables

ijuma opened the pull request:
There should be only one cases where these clean-ups have a functional impact: replaced repeated identical logs with a single log for the stale controller epoch case.

The rest should just make the code easier to read and make it a bit less wasteful. I did this exercise because unused variables sometimes mask bugs.

ijuma committed:
Remove unused fields, variables and methods

ijuma committed:
Remove unnecessary initialisation

ijuma committed:
Use underscore instead of unused variables

ijuma committed:
Use `keys`, `values` and similar methods to avoid unused variables

ijuma committed:
Avoid unnecessary `toSet` and replace pattern matching by `isDefined`

ijuma committed:
Replace one log entry per partition with a single log entry for a stale controller epoch error

Since the same information was printed each time, it seems completely pointless to print it
multiple times.

ijuma committed:
Controller clean-ups related to unused variables

ijuma commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -639,11 +638,9 @@ class ReplicaManager(val config: KafkaConfig,
         stateChangeLogger.warn(("Broker %d ignoring LeaderAndIsr request from controller %d with correlation id %d since " +
           "its controller epoch %d is old. Latest known controller epoch is %d").format(localBrokerId, leaderAndISRRequest.controllerId,
           correlationId, leaderAndISRRequest.controllerEpoch, controllerEpoch))
-        }
```
This is the case where we used to log the exact same identical line for each partition.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -65,9 +65,6 @@
     private Set<String> subscription;
 
-    /* the list of partitions the user has requested */
-    private Set<TopicPartition> userAssignment;
```
A bit surprised about this one.

hachikuji replied:
Ouch. Not sure how I missed that.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -608,7 +608,7 @@ object AdminUtils extends Logging with AdminUtilities {
     def entityPaths(zkUtils: ZkUtils, rootPath: Option[String]): Seq[String] = {
       val root = rootPath match {
-        case Some(path) => rootEntityType + '/' + rootPath
+        case Some(path) => rootEntityType + '/' + path
```
@rajinisivaram, can you please check if this is correct? If so, do you know why our tests didn't catch it? How could we improve the tests so that we can verify the fix?

rajinisivaram replied:
@ijuma Oops, will add a test. Would it be better if I created a PR with this change and tests? It will be good to apply the change to the 0.10.1 branch as well. Thank you for spotting it.

rajinisivaram replied:
@ijuma Actually, the code above doesn't have any impact unless someone goes and explicitly deletes config change notifications from Zookeeper. I can't find any code that deletes change notification znodes from Zookeeper - do you know if that is intentional?

ijuma replied:
Thanks. I don't know, but we can ask Jun. Let's do it in https://github.com/apache/kafka/pull/1987 since it's more focused.

ijuma commented:
@hachikuji, would you be able to review this perhaps?

ijuma committed:
Fix `IsrExpirationTest` log mock

Believe it or not, this started failing after we removed a
call whose result was unused.

ijuma commented on core/src/test/scala/unit/kafka/server/ISRExpirationTest.scala:
```diff
@@ -66,7 +66,7 @@ class IsrExpirationTest {
    */
   @Test
   def testIsrExpirationForStuckFollowers() {
```
This and the other test started failing because the following line was removed:

https://github.com/apache/kafka/pull/1985/files#diff-5450baca03f57b9f2030f93a480e6969L418

Got to love mocks. :)

ijuma commented:
Rebased on trunk after merging #1987 and edited PR description to take that into account.

ijuma committed:
Fix edge case in `MiniKdc` uncovered by unused variable

ijuma committed:
Simplify `FetcherFetch.fetch` logic

ijuma committed:
Pass the created `dirs` to `createLogManager`

hachikuji commented:
Thanks @ijuma, Assuming tests are passing, LGTM. Kudos to anyone who can remove over 200 lines from the codebase!

hachikuji commented:
@ijuma Not sure I've seen the above test failure before:

```
java.lang.AssertionError: expected:<Message(magic = 1, attributes = 0, CreateTime = -1, crc = 3573431224, key = null, payload = java.nio.HeapByteBuffer[pos=0 lim=4 cap=4])> but was:<Message(magic = 1, attributes = 0, CreateTime = 1477319907056, crc = 2050317473, key = java.nio.HeapByteBuffer[pos=0 lim=4 cap=12], payload = java.nio.HeapByteBuffer[pos=0 lim=4 cap=4])>
```

ijuma commented:
Will check, thanks.

ijuma committed:
Fix broken `testAsyncSendCanCorrectlyFailWithTimeout`

It previously was passing because we were catching all exceptions
being thrown unintentionally.

ijuma commented:
Fixed the failure, checked with Jason that the fix was OK. Subsequent Jenkins failure was a known transient failure. Tests passed locally so merged to trunk.
